### DISCUSSION TITLE
`ValidationException` from addons doesn't get caught by forms AJAX handling. Is this intentional?

### DISCUSSION THREAD

**Discussion by sjclark** [statamic/cms#10375] [Troubleshooting]
😄 An unusual issue I stumbled across after having a chat with @martyf, @robdekort and @duncanmcclean - figured it was worth starting a discussion!

As far as I can tell many Statamic addons that throw form errors all throw a `ValidationException::withMessages`: eg;
```php
throw ValidationException::withMessages([__('statamic-turnstile::validation.turnstile')]);
##
throw ValidationException::withMessages([config('recaptcha.recaptcha_v3.error_message')]);
```
This works well for _normal_ Statamic forms, but is incompatible with the Statamic AJAX error responses, as these are made before the addon code exception is thrown. This can result in unusual (or broken?) forms, or errors being turned in different JSON responses:

![image](https://github.com/statamic/cms/assets/4531017/5ec8e7ac-91be-47a1-944d-ac42b8d17bc4)
(normal error response)
![image](https://github.com/statamic/cms/assets/4531017/28e2bf8c-b302-4bd2-80d5-98b86bcf2eb2)
(addon error response)

This is all slightly above my pay grade, but from my understanding Statamic responds to the AJAX form requests in a fairly Statamic way, as opposed to a Laravel way. I was wondering if there was some sort of way that this gap could be bridged? Or failing that if there was any way to adapt the addons to be able to hook in? 🤷‍♂️ 

```php
## Statamics current FormController response
    private function formFailure($params, $errors, $form)
    {
        if (request()->ajax() || request()->wantsJson()) {
            return response([
                'errors' => (new MessageBag($errors))->all(),
                'error' => collect($errors)->map(function ($errors, $field) {
                    return $errors[0];
                })->all(),
            ], 400);
        }
        ```

**Comment by ryanmitchell** [comment]
What is the actual issue this is causing for you? can you share code that shows it?

**Reply by sjclark** [reply]
Essentially if you add any of the addons that add any sort of front-end form validations then these will not correctly throw errors; a good example is https://statamic.com/addons/aryeh-raber/captcha

This addon (like others) throws an error that is correctly interpreted by normal Statamic forms, but as soon as you use AJAX forms then errors are sent through in the JSON format above, which is not compatible with the normal json errors object response.

My front-end AJAX code is functionally just a copy of Statamic Peaks form handler.

Unfortunately I had to code around this in the project where I was exploring this (as it had to launch today), but I can get a minimal case together in a couple of days if neccessary.

**Reply by martyf** [reply]
I think another word to say it is - with AJAX form submissions of course:

Statamic returns an `errors` _object_ with a key/value pair of field and error. Peak (and other kits) use this. So validation and precognition work nicely. 

But addons throw an exception - as you would in Laravel - however the response is an `errors` _array_. No key/value pair in the `errors` variable. 

So while Statamic returns one structure, Laravel returns another. So it means any front end JS needs to look for both to function fully.

**Reply by ryanmitchell** [reply]
Thanks, I see the issue

**Reply by ryanmitchell** [reply]
https://github.com/statamic/cms/pull/10376

**Reply by sjclark** [reply]
> #10376

😄  amazing, it's late here but I'll try have a look at this tomorrow (or day after)!
